tensorstore/kvstore/s3/s3_credential_provider.cc

```diff
@@ -0,0 +1,245 @@
+      return absl::NotFoundError("Could not read $HOME");
+    }
+    result = JoinPath(*home_dir, kDefaultAwsDirectory, kDefaultAwsCredentialsFile);
+  } else {
+    result = *credentials_file;
+  }
+  if(!IsFile(result)) {
+    return absl::NotFoundError(
+      absl::StrCat("Could not find the credentials file at "
+                   "location [", result, "]"));
+  }
+  return result;
+}
+
+
+Result<std::unique_ptr<CredentialProvider>> GetDefaultS3CredentialProvider(
+    std::string_view profile,
+    std::shared_ptr<internal_http::HttpTransport> transport) {
+
+  // 1. Obtain credentials from environment variables
+  if(auto access_key = GetEnv(kEnvAwsAccessKeyId); access_key) {
+    ABSL_LOG(INFO) << "Using Environment Variable S3CredentialProvider";
+    S3Credentials credentials;
+    credentials.access_key = *access_key;
+    auto secret_key = GetEnv(kEnvAwsSecretAccessKey);
+
+    if(secret_key.has_value()) {
+      credentials.secret_key = *secret_key;
+    }
+
+    auto session_token = GetEnv(kEnvAwsSessionToken);
+
+    if(session_token.has_value()) {
+      credentials.session_token = *session_token;
+    }
+
+    return std::make_unique<EnvironmentCredentialProvider>(std::move(credentials));
+  }
+
+  // 2. Obtain credentials from AWS_SHARED_CREDENTIALS_FILE or ~/.aws/credentials
+  if(auto credentials_file = GetS3CredentialsFileName(); credentials_file) {
+    std::optional<std::string> env_profile;  // value must not outlive view
+    if(profile.empty()) {
+        env_profile = GetEnv(kEnvAwsDefaultProfile);
+        if(!env_profile) env_profile = GetEnv(kEnvAwsProfile);
+        profile = !env_profile ? kDefaultProfile : *env_profile;
+    }
+    ABSL_LOG(INFO) << "Using File S3CredentialProvider with profile " << profile;
+    return std::make_unique<FileCredentialProvider>(credentials_file.value(), profile);
+  }
+
+  // 3. Obtain credentials from EC2 Metadata server
+  if(false) {
+    ABSL_LOG(INFO) << "Using EC2 Metadata Service S3CredentialProvider";
+    return std::make_unique<EC2MetadataCredentialProvider>(transport);
+  }
+
+  return absl::NotFoundError(
+    "No credentials provided in environment variables, "
+    "credentials file not found and not running on AWS.");
+}
+
+struct CredentialProviderRegistry {
+ std::vector<std::pair<int, S3CredentialProvider>> providers;
+ absl::Mutex mutex;
+};
+
+CredentialProviderRegistry& GetS3ProviderRegistry() {
+ static internal::NoDestructor<CredentialProviderRegistry> registry;
+ return *registry;
+}
+
+} // namespace
+
+/// https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-format
+Result<S3Credentials> FileCredentialProvider::GetCredentials() {
+  absl::ReaderMutexLock lock(&mutex_);
+  std::ifstream ifs(filename_);
+
+  if (!ifs) {
+    return absl::NotFoundError(
+        absl::StrCat("Could not open the credentials file [", filename_, "]"));
+  }
+
+  S3Credentials credentials;
+  std::string section_name;
```

Comment:
Use std::string_view for section_name.

Reply:
We need a `std::string` here because `section_name` would end up referencing `line`, which changes on each iteration of the loop.